pybats IdlFile sort is nondeterministic

The CI is failing on Python 3.9 and 3.10 with numpy 1.25. This is in TestIdlFile.testBinary and the check that fails is the comparison to knownMhdX_sorted.

The problem is at the end of `_read_idl_bin`, where the sorting is done on an array with many repeated elements (`gridtotal`)--there are 2560 elements but only 167 are unique. So this is heavily dependent on the specific sort algorithm, and the default numpy algorithm is _not_ stable. It's entirely possible the expected values are wrong! In fact, specifying a stable algorithm makes the test fail on versions of numpy where it used to pass (e.g. Python 3.9, numpy 1.18.5).

Here's the code:
```python
gridtotal = np.zeros(npts)
offset = 0.0  # The offset ensures no repeating vals while sorting.
for key in pbdat['grid'].attrs['dims']:
    gridtotal = gridtotal + offset + pbdat[key]
    offset = offset + np.pi/2.0
    SortIndex = np.argsort(gridtotal)
for key in list(names[:nvar+ndim]):
    if key == 'grid': continue
    pbdat[key] = pbdat[key][SortIndex]
```

This pretty clearly refactors to:
```python
gridtotal = np.zeros(npts)
for key in pbdat['grid'].attrs['dims']:
    gridtotal += pbdat[key]
dimcount = len(pbdat['grid'].attrs['dims'])
gridtotal += np.pi / 2.0 * (dimcount - 1) * dimcount / 2
for key in list(names[:nvar+ndim]):
    if key == 'grid': continue
    pbdat[key] = pbdat[key][SortIndex]
```
i.e. there's only one sort, and the offset does nothing to guarantee uniqueness.

The definition of the intended behavior ("nearby positions are arranged as close as possible to each other in the data arrays") is fairly vague so it's hard to tell what's really desired here.

### Error message/Traceback:

```
======================================================================
FAIL: testBinary (__main__.TestIdlFile)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jnik/work/scm/spacepy/tests/test_pybats.py", line 294, in testBinary
    numpy.testing.assert_array_equal(actual, self.knownMhdX_sorted)
  File "/home/jnik/miniconda/envs/py310/lib/python3.10/site-packages/numpy/testing/_private/utils.py", line 924, in assert_array_equal
    assert_array_compare(operator.__eq__, x, y, err_msg=err_msg,
  File "/home/jnik/miniconda/envs/py310/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/home/jnik/miniconda/envs/py310/lib/python3.10/site-packages/numpy/testing/_private/utils.py", line 801, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not equal

Mismatched elements: 4 / 8 (50%)
Max absolute difference: 16.
Max relative difference: 0.07843137
 x: array([-220., -212., -220., -220., -204., -212., -196., -212.],
      dtype=float32)
 y: array([-220., -212., -220., -204., -220., -212., -212., -204.])

----------------------------------------------------------------------
Ran 1 test in 0.004s

FAILED (failures=1)
```

### OS, Python version, and dependency version information
```
Linux-6.2.0-24-generic-x86_64-with-glibc2.37
sys.version_info(major=3, minor=10, micro=12, releaselevel='final', serial=0)
numpy=1.25.0
scipy=1.11.1
matplotlib=3.7.1
```


### Version of SpacePy
Current github main (d1569c476e6bd4c2bd760e30e9a7abaf006942f6)

--- a/Doc/source/release_notes.rst
+++ b/Doc/source/release_notes.rst
@@ -63,11 +63,6 @@ with temperature and density naming.
 perpendicular velocities with new `~.pybats.bats.Bats2d.calc_uperp`
 and `~.pybats.bats.Bats2d.calc_upar` methods.
 
-`~.pybats.IdlFile` supports reading binary files without sorting
-unstructured data (and adds support for sorting in reading ASCII
-files). Unsorted is now the default for all except `~.pybats.bats.Bats2d`.
-Thanks Lutz Rastaetter.
-
 Deprecations and removals
 *************************
 Since plot styles are no longer applied on import, importing
@@ -80,6 +75,9 @@ been fixed.
 
 Other changes
 *************
+`~.pybats.IdlFile` no longer sorts unstructured data from binary files; see
+that documentation for details. Thanks Lutz Rastaetter.
+
 :func:`~.datamodel.toCDF` now only accepts valid keyword arguments.
 
 Plot styles are not automatically applied on import of :mod:`.plot`. Use

--- a/spacepy/pybats/__init__.py
+++ b/spacepy/pybats/__init__.py
@@ -400,7 +400,7 @@ def add_body(ax, rad=2.5, facecolor='lightgrey', show_planet=True,
     ax.add_artist(body)
 
 
-def _read_idl_ascii(pbdat, header='units', start_loc=0, keep_case=True, sort_unstructured=False):
+def _read_idl_ascii(pbdat, header='units', start_loc=0, keep_case=True):
     '''
     Load a SWMF IDL ascii output file and load into a pre-existing PbData
     object.  This should only be called by :class:`IdlFile`.
@@ -547,18 +547,7 @@ def _read_idl_ascii(pbdat, header='units', start_loc=0, keep_case=True, sort_uns
                 pbdat[v] = dmarray(np.reshape(pbdat[v], pbdat['grid'],
                                               order='F'), attrs=pbdat[v].attrs)
 
-    # Unstructured data can be in any order, so let's sort it.
-    if gtyp == 'Unstructured' and sort_unstructured:
-        gridtotal = np.zeros(npts)
-        offset = 0.0  # The offset ensures no repeating vals while sorting.
-        for key in pbdat['grid'].attrs['dims']:
-            gridtotal = gridtotal + offset + pbdat[key]
-            offset = offset + np.pi/2.0
-            SortIndex = np.argsort(gridtotal)
-        for key in list(pbdat.keys()):
-            if key=='grid': continue
-            pbdat[key] = pbdat[key][SortIndex]
-                
+
 def readarray(f, dtype=np.float32, inttype=np.int32):
     '''
     Read an array from an unformatted binary file written out by a
@@ -839,9 +828,8 @@ def _probe_idlfile(filename):
 
 
 def _read_idl_bin(pbdat, header='units', start_loc=0, keep_case=True,
-                  headeronly=False, sort_unstructured=False):
-    '''
-    Load a SWMF IDL binary output file and load into a pre-existing PbData
+                  headeronly=False):
+    '''Load a SWMF IDL binary output file and load into a pre-existing PbData
     object.  This should only be called by :class:`IdlFile`, which will
     include information on endianess and size of integers & floating point
     values.
@@ -855,6 +843,19 @@ def _read_idl_bin(pbdat, header='units', start_loc=0, keep_case=True,
     assumed to be a list of units for each variable contained within the
     file.  If set to **None** or not recognized, the header will be saved
     in the object's attribute list under 'header'.
+
+    .. versionchanged:: 0.5.0
+
+       Unstructured data are presented as in the files. When reading
+       3D magnetosphere files, this preserves the 3D block structure,
+       as required for the BATSRUS interpolator in the `Kamodo
+       Heliophysics model readers package
+       <https://github.com/nasa/kamodo>`_. Before 0.5.0, binary
+       unstructured data were sorted in an attempt to put nearby
+       positions close to each other in the data arrays. This sorting
+       was nondeterministic and has been removed; see
+       `bats.Bats2d.extract` and `qotree.QTree` for processing
+       adjacent cells.
 
     Parameters
     ----------
@@ -874,14 +875,6 @@ def _read_idl_bin(pbdat, header='units', start_loc=0, keep_case=True,
     keep_case : boolean
         If set to True, the case of variable names will be preserved.  If
         set to False, variable names will be set to all lower case.
-
-    sort_unstructured : bool, default False
-
-      .. versionadded:: 0.5.0
-
-      In unstructured data, sort so nearby positions are arranged as close
-      as possible to each other in the data arrays. (This was the behavior
-      before 0.5.0).
 
     Returns
     -------
@@ -1012,18 +1005,6 @@ def _read_idl_bin(pbdat, header='units', start_loc=0, keep_case=True,
                 # Put data into multidimensional arrays.
                 pbdat[names[i]] = pbdat[names[i]].reshape(
                     pbdat['grid'], order='F')
-
-        # Unstructured data can be in any order, so let's sort it.
-        if gtyp == 'Unstructured' and sort_unstructured:
-            gridtotal = np.zeros(npts)
-            offset = 0.0  # The offset ensures no repeating vals while sorting.
-            for key in pbdat['grid'].attrs['dims']:
-                gridtotal = gridtotal + offset + pbdat[key]
-                offset = offset + np.pi/2.0
-                SortIndex = np.argsort(gridtotal)
-            for key in list(names[:nvar+ndim]):
-                if key == 'grid': continue
-                pbdat[key] = pbdat[key][SortIndex]
 
 
 class PbData(SpaceData):
@@ -1182,6 +1163,20 @@ class IdlFile(PbData):
     doesn't work, the error will manifest itself through the "struct" package
     as an "unpack requires a string of argument length 'X'".
 
+    .. versionchanged:: 0.5.0
+
+       Unstructured data are presented as in the files. When reading
+       3D magnetosphere files, this preserves the 3D block structure,
+       as required for the BATSRUS interpolator in the `Kamodo
+       Heliophysics model readers package
+       <https://github.com/nasa/kamodo>`_. Before 0.5.0, binary
+       unstructured data were sorted in an attempt to put nearby
+       positions close to each other in the data arrays. This sorting
+       was nondeterministic and has been removed; see
+       :meth:`~spacepy.pybats.bats.Bats2d.extract` and
+       :class:`~spacepy.pybats.qotree.QTree` for processing
+       adjacent cells. (ASCII data were never sorted.)
+
     Parameters
     ----------
     filename : string
@@ -1196,23 +1191,10 @@ class IdlFile(PbData):
     keep_case : bool
         If set to True, the case of variable names will be preserved.  If
         set to False, variable names will be set to all lower case.
-
-    sort_unstructured : bool, default: ``False``
-
-        .. versionadded:: 0.5.0
-
-        If ``True``, the positions and data are sorted so nearby
-        positions are arranged as close as possible to each other in
-        the data arrays. If ``False`` (default) the positions in the
-        unstructured grid data are not rearranged. When reading 3D
-        magnetosphere files, this preserves the 3D block structure, as
-        required for the BATSRUS interpolator in the `Kamodo Heliophysics
-        model readers package <https://github.com/nasa/kamodo>`_.
-        (Subclasses may have different defaults.)
     '''
 
     def __init__(self, filename, iframe=0, header='units',
-                 keep_case=True, sort_unstructured=False, *args, **kwargs):
+                 keep_case=True, *args, **kwargs):
         super(IdlFile, self).__init__(*args, **kwargs)  # Init as PbData.
 
         # Gather information about the file: format, endianess (if necessary),
@@ -1242,7 +1224,7 @@ class IdlFile(PbData):
             self._scan_asc_frames()
 
         # Read one entry of the file (defaults to first frame):
-        self.read(iframe=iframe, sort_unstructured=sort_unstructured)
+        self.read(iframe=iframe)
 
         # Update information about the currently loaded frame.
         self.attrs['iframe'] = iframe
@@ -1360,16 +1342,11 @@ class IdlFile(PbData):
     def __repr__(self):
         return 'SWMF IDL-Binary file "%s"' % (self.attrs['file'])
 
-    def read(self, iframe=0, sort_unstructured=False):
+    def read(self, iframe=0):
         '''
         This method reads an IDL-formatted BATS-R-US output file and places
         the data into the object.  The file read is self.filename which is
         set when the object is instantiated.
-
-        .. versionchanged:: 0.5.0
-
-        The new sort_unstructured keyword can enable data sorting
-        (formerly default in the binary file reader).
         '''
 
         # Get location of frame that we wish to read:
@@ -1377,12 +1354,10 @@ class IdlFile(PbData):
 
         if self.attrs['format'] == 'asc':
             _read_idl_ascii(self, header=self._header, start_loc=loc,
-                            keep_case=self._keep_case,
-                            sort_unstructured=sort_unstructured)
+                            keep_case=self._keep_case)
         elif self.attrs['format'] == 'bin':
             _read_idl_bin(self, header=self._header, start_loc=loc,
-                          keep_case=self._keep_case,
-                          sort_unstructured=sort_unstructured)
+                          keep_case=self._keep_case)
         else:
             raise ValueError('Unrecognized file format: {}'.format(
                 self._format))

--- a/spacepy/pybats/bats.py
+++ b/spacepy/pybats/bats.py
@@ -753,16 +753,19 @@ class Bats2d(IdlFile):
     see how to handle multi-frame files (*.outs) and for a list of critical
     attributes.
 
+    .. versionchanged:: 0.5.0
+
+       Unstructured data are now presented as in the files. See
+       `~pybats.IdlFile` for details.
     '''
     # Init by calling IdlFile init and then building qotree, etc.
-    def __init__(self, filename, *args, sort_unstructured=True, **kwargs):
+    def __init__(self, filename, *args, **kwargs):
 
         # Create quad tree object attribute:
         self._qtree = None
 
         # Read file.
         IdlFile.__init__(self, filename, keep_case=False, *args,
-                         sort_unstructured=sort_unstructured,
                          **kwargs)
 
         # Behavior of output files changed Jan. 2017:

--- a/tests/test_pybats.py
+++ b/tests/test_pybats.py
@@ -244,7 +244,6 @@ class TestIdlFile(unittest.TestCase):
     knownMhdZlim = 124.0
     knownMhdTime = dt.datetime(2014, 4, 10, 0, 0, 50)
     knownMhdX_unsorted = [-220., -212., -204., -196., -188., -180., -172., -164.] # first 8 X positions in the y0_ascii.out file
-    knownMhdX_sorted = [-220., -212., -220., -204., -220., -212., -212., -204.] # first 8 X positions in the binary y=0_mhd_1_e20140410-000050.out file after sorting of positions
 
     # Known values for multi-frame *.outs files:
     # Time/iteration range covered by files:
@@ -286,12 +285,6 @@ class TestIdlFile(unittest.TestCase):
         self.assertEqual(self.knownMhdZlim*-1, mhd['z'].min())
         for v in range(len(self.knownMhdX_unsorted)):
             self.assertEqual(self.knownMhdX_unsorted[v], (mhd['x'])[v])
-        mhd = pb.IdlFile(os.path.join(spacepy_testing.datadir,
-                                      'pybats_test',
-                                      'y=0_mhd_1_e20140410-000050.out'),
-                         sort_unstructured=True)
-        for v in range(len(self.knownMhdX_sorted)):
-            self.assertEqual(self.knownMhdX_sorted[v], (mhd['x'])[v])
 
     def testAscii(self):
         # Open file:

